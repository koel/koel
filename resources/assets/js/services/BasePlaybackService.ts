import isMobile from 'ismobilejs'
import { throttle } from 'lodash'
import plyr from 'plyr'
import { watch } from 'vue'
import { volumeManager } from '@/services/volumeManager'

export abstract class BasePlaybackService {
  public player!: Plyr
  private boundMediaEvents = new Set<[string, EventListener, boolean]>()

  public activate (plyrWrapper: HTMLElement) {
    if (!this.player) {
      // player and watchers can be kept between init/destroy sessions
      this.player = this.player || plyr.setup(plyrWrapper, { controls: [] })[0]
      watch(volumeManager.volume, volume => this.player.setVolume(volume), { immediate: true })
      this.setMediaSessionActionHandlers()
    }

    if (!this.boundMediaEvents.size) {
      this.addMediaEventListeners()
    }

    return this
  }

  protected setMediaSessionActionHandlers () {
    if (!navigator.mediaSession) {
      return
    }

    navigator.mediaSession.setActionHandler('play', () => this.resume())
    navigator.mediaSession.setActionHandler('pause', () => this.pause())
    navigator.mediaSession.setActionHandler('stop', () => this.stop())
    navigator.mediaSession.setActionHandler('previoustrack', () => this.playPrev())
    navigator.mediaSession.setActionHandler('nexttrack', () => this.playNext())

    if (!isMobile.apple) {
      navigator.mediaSession.setActionHandler('seekbackward', details => this.rewind(details.seekOffset || 10))
      navigator.mediaSession.setActionHandler('seekforward', details => this.forward(details.seekOffset || 10))
    }

    navigator.mediaSession.setActionHandler('seekto', details => {
      if (details.fastSeek && 'fastSeek' in this.player.media) {
        this.fastSeek(details.seekTime || 0)
      } else {
        this.seekTo(details.seekTime || 0)
      }
    })
  }

  private addMediaEventListeners () {
    const listen = (event: string, handler: Closure, options?: boolean | AddEventListenerOptions) => {
      this.player.media.addEventListener(event, handler, options)
      this.boundMediaEvents.add([event, handler, !!options])
    }

    listen('error', this.onError.bind(this), true)
    listen('ended', this.onEnded.bind(this))

    const timeUpdateHandler = process.env.NODE_ENV === 'test' ? this.onTimeUpdate : throttle(this.onTimeUpdate, 1000)
    listen('timeupdate', timeUpdateHandler.bind(this))
  }

  public abstract play (source: Streamable): Promise<void>

  public abstract stop (): Promise<void>

  public abstract pause (): Promise<void>

  public abstract resume (): Promise<void>

  public abstract playNext (): Promise<void>

  public abstract playPrev (): Promise<void>

  public abstract toggle (): Promise<void>

  public abstract rewind (seconds: number): void

  public abstract forward (seconds: number): void

  public abstract seekTo (position: number): void

  public abstract fastSeek (position: number): void

  protected abstract onError (error: ErrorEvent): void

  protected abstract onEnded (event: Event): void

  protected abstract onTimeUpdate (event: Event): void

  public abstract rotateRepeatMode (): void

  public deactivate () {
    // Upon deactivating (i.e. switching to another playback service) we want to remove all event listeners
    // to not mess up the media events in the other service.
    this.boundMediaEvents.forEach(([event, handler, options]) => {
      this.player.media.removeEventListener(event, handler, options)
    })

    this.boundMediaEvents.clear()

    this.stop()
  }
}
