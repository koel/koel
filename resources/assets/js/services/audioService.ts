import { equalizerStore } from '@/stores'
import { frequencies } from '@/config'
import { dbToGain } from '@/utils'

interface Band {
  label: string
  filter: BiquadFilterNode
  db: number
}

export const audioService = {
  unlocked: false,

  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,
  preampGainNode: null as unknown as GainNode,
  analyzer: null as unknown as AnalyserNode,

  bands: [] as Band[],

  init (mediaElement: HTMLMediaElement) {
    this.element = mediaElement

    this.context = new AudioContext()
    this.preampGainNode = this.context.createGain()
    this.source = this.context.createMediaElementSource(this.element)
    this.analyzer = this.context.createAnalyser()

    this.source.connect(this.preampGainNode)

    const config = equalizerStore.getConfig()

    this.changePreampGain(config.preamp)

    let prevFilter: BiquadFilterNode

    // Create 10 bands with the frequencies similar to those of Winamp and connect them together.
    frequencies.forEach((frequency, i) => {
      const filter = this.context.createBiquadFilter()

      if (i === 0) {
        filter.type = 'lowshelf'
      } else if (i === frequencies.length - 1) {
        filter.type = 'highshelf'
      } else {
        filter.type = 'peaking'
      }

      filter.Q.setTargetAtTime(1, this.context.currentTime, 0.01)
      filter.frequency.setTargetAtTime(frequency, this.context.currentTime, 0.01)
      filter.gain.value = dbToGain(config.gains[i])

      prevFilter ? prevFilter.connect(filter) : this.preampGainNode.connect(filter)
      prevFilter = filter

      this.bands.push({
        filter,
        label: String(frequency).replace('000', 'K'),
        db: config.gains[i]
      })
    })

    prevFilter!.connect(this.analyzer)

    // connect the analyzer node last, so that changes to the equalizer affect the visualizer as well
    this.analyzer.connect(this.context.destination)

    this.unlockAudioContext()
  },

  changePreampGain (db: number) {
    this.preampGainNode.gain.value = dbToGain(db)
  },

  changeFilterGain (node: BiquadFilterNode, db: number) {
    this.bands.find(band => band.filter === node)!.db = db
    node.gain.value = dbToGain(db)
  },

  /**
   * Attempt to unlock the audio context on mobile devices by creating and playing a silent buffer upon the
   * first user interaction.
   */
  unlockAudioContext () {
    ['touchend', 'touchstart', 'click'].forEach(event => {
      document.addEventListener(event, () => {
        if (this.unlocked) return

        const source = this.context.createBufferSource()
        source.buffer = this.context.createBuffer(1, 1, 22050)
        source.connect(this.context.destination)
        source.start(0)

        this.unlocked = true
      }, {
        once: true
      })
    })
  }
}
