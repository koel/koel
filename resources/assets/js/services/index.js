export * from "./info";
export * from "./download";
export * from "./http";
export * from "./ls";
export * from "./playback";
export * from "./youtube";
export * from "./socket";
