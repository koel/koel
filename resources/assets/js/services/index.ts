export * from './http'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
export * from './mediaInfoService'
export * from './cache'
export * from './socketListener'
export * from './volumeManager'
