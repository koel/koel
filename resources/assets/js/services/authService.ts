import { localStorageService } from '@/services'

const API_TOKEN_STORAGE_KEY = 'api-token'
const AUDIO_TOKEN_STORAGE_KEY = 'audio-token'

export const authService = {
  getApiToken: () => localStorageService.get(API_TOKEN_STORAGE_KEY),

  hasApiToken () {
    return Boolean(this.getApiToken())
  },

  setApiToken: (token: string) => localStorageService.set(API_TOKEN_STORAGE_KEY, token),

  destroy: () => {
    localStorageService.remove(API_TOKEN_STORAGE_KEY)
    localStorageService.remove(AUDIO_TOKEN_STORAGE_KEY)
  },

  setAudioToken: (token: string) => localStorageService.set(AUDIO_TOKEN_STORAGE_KEY, token),

  getAudioToken: () => {
    // for backward compatibility, we first try to get the audio token, and fall back to the (full-privileged) API token
    return localStorageService.get(AUDIO_TOKEN_STORAGE_KEY) || localStorageService.get(API_TOKEN_STORAGE_KEY)
  }
}
