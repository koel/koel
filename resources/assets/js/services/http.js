import axios from "axios";
import NProgress from "nprogress";

import { event } from "@/utils";
import { ls } from "@/services";

/**
 * Responsible for all HTTP requests.
 */
export const http = {
  request(method, url, data, successCb = null, errorCb = null) {
    axios
      .request({
        url,
        data,
        method: method.toLowerCase()
      })
      .then(successCb)
      .catch(errorCb);
  },

  get(url, successCb = null, errorCb = null) {
    return this.request("get", url, {}, successCb, errorCb);
  },

  post(url, data, successCb = null, errorCb = null) {
    return this.request("post", url, data, successCb, errorCb);
  },

  put(url, data, successCb = null, errorCb = null) {
    return this.request("put", url, data, successCb, errorCb);
  },

  delete(url, data = {}, successCb = null, errorCb = null) {
    return this.request("delete", url, data, successCb, errorCb);
  },

  /**
   * Init the service.
   */
  init() {
    axios.defaults.baseURL = "/api";

    // Intercept the request to make sure the token is injected into the header.
    axios.interceptors.request.use(config => {
      config.headers.Authorization = `Bearer ${ls.get("jwt-token")}`;
      return config;
    });

    // Intercept the response and…
    axios.interceptors.response.use(
      response => {
        NProgress.done();

        // …get the token from the header or response data if exists, and save it.
        const token =
          response.headers["Authorization"] || response.data["token"];
        if (token) {
          ls.set("jwt-token", token);
        }

        return response;
      },
      error => {
        NProgress.done();
        // Also, if we receive a Bad Request / Unauthorized error
        if (error.response.status === 400 || error.response.status === 401) {
          // and we're not trying to login
          if (
            !(
              error.config.method === "post" &&
              /\/api\/me\/?$/.test(error.config.url)
            )
          ) {
            // the token must have expired. Log out.
            event.emit("logout");
          }
        }

        return Promise.reject(error);
      }
    );
  }
};
