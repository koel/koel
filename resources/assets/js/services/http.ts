import Axios, { AxiosInstance, Method } from 'axios'
import NProgress from 'nprogress'
import { eventBus } from '@/utils'
import { authService } from '@/services'

class Http {
  client: AxiosInstance

  private static setProgressBar () {
    NProgress.start()
  }

  private static hideProgressBar () {
    NProgress.done(true)
  }

  public request<T> (method: Method, url: string, data: Record<string, any> = {}, onUploadProgress?: any) {
    return this.client.request({
      url,
      data,
      method,
      onUploadProgress
    }) as Promise<{ data: T }>
  }

  public async get<T> (url: string) {
    return (await this.request<T>('get', url)).data
  }

  public async post<T> (url: string, data: Record<string, any>, onUploadProgress?: any) {
    return (await this.request<T>('post', url, data, onUploadProgress)).data
  }

  public async put<T> (url: string, data: Record<string, any>) {
    return (await this.request<T>('put', url, data)).data
  }

  public async delete<T> (url: string, data: Record<string, any> = {}) {
    return (await this.request<T>('delete', url, data)).data
  }

  constructor () {
    this.client = Axios.create({
      baseURL: `${window.BASE_URL}api`,
      headers: {
        'X-Api-Version': 'v6'
      }
    })

    // Intercept the request to make sure the token is injected into the header.
    this.client.interceptors.request.use(config => {
      Http.setProgressBar()
      config.headers.Authorization = `Bearer ${authService.getApiToken()}`
      return config
    })

    // Intercept the response and…
    this.client.interceptors.response.use(response => {
      Http.hideProgressBar()

      // …get the token from the header or response data if exists, and save it.
      const token = response.headers.authorization || response.data.token
      token && authService.setApiToken(token)

      const audioToken = response.data['audio-token']
      audioToken && authService.setAudioToken(audioToken)

      return response
    }, error => {
      Http.hideProgressBar()

      // Also, if we receive a Bad Request / Unauthorized error
      if (error.response?.status === 400 || error.response?.status === 401) {
        // and we're not trying to log in
        if (!(error.config.method === 'post' && /\/api\/me\/?$/.test(error.config.url))) {
          // the token must have expired. Log out.
          eventBus.emit('LOG_OUT')
        }
      }

      return Promise.reject(error)
    })
  }
}

export const http = new Http()
