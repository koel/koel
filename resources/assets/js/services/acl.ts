import { http } from '@/services/http'
import { cache } from '@/services/cache'

type ResourceType = 'album' | 'artist' | 'radio-station' | 'user'
type Action = 'edit' | 'delete'

export const acl = {
  checkResourcePermission: async (type: ResourceType, id: string | number, action: Action) => {
    return await cache.remember(['permission', type, id, action], async () => {
      const { allowed } = await http.silently.get<{ allowed: boolean }>(`acl/permissions/${type}/${id}/${action}`)

      return allowed
    })
  },

  fetchAssignableRoles: async () => {
    return await cache.remember(['assignable-roles'], async () => {
      const { roles } = await http.get<{
        roles: Array<{ id: Role, label: string, description: string }>
      }>('acl/assignable-roles')

      return roles
    })
  },
}
