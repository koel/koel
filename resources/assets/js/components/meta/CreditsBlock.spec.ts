import { describe, expect, it } from 'vitest'
import { createHarness } from '@/__tests__/TestHarness'
import { http } from '@/services/http'
import Component from './CreditsBlock.vue'

describe('creditsBlock.vue', () => {
  const h = createHarness()

  it('renders the credits', async () => {
    window.IS_DEMO = true

    const getMock = h.mock(http, 'get').mockResolvedValue([
      { name: 'Foo', url: 'https://foo.com' },
      { name: 'Bar', url: 'https://bar.com' },
      { name: 'Something Else', url: 'https://something-else.net' },
    ])

    const { html } = h.render(Component)

    await h.tick(3)
    expect(html()).toMatchSnapshot()
    expect(getMock).toHaveBeenCalledWith('demo/credits')

    window.IS_DEMO = false
  })
})
