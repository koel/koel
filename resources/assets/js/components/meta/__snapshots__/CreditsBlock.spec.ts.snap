// Vitest Snapshot v1

exports[`creditsBlock.vue > renders the credits 1`] = `
<div data-v-7d3b4fcd="" class="max-h-[9rem] overflow-auto" data-testid="demo-credits"> Music by <ul data-v-7d3b4fcd="" class="inline">
    <li data-v-7d3b4fcd="" class="inline"><a data-v-7d3b4fcd="" href="https://bar.com" target="_blank">Bar</a></li>
    <li data-v-7d3b4fcd="" class="inline"><a data-v-7d3b4fcd="" href="https://foo.com" target="_blank">Foo</a></li>
    <li data-v-7d3b4fcd="" class="inline"><a data-v-7d3b4fcd="" href="https://something-else.net" target="_blank">Something Else</a></li>
  </ul>
</div>
`;
